### ISSUE TITLE
Support linter configuration through a JSON file

### ISSUE THREAD

**Issue by steelbrain** [atom-community/linter#744]
The JSON file would be in the root of the project dir and would have some basic on/off features. Like
```
{
  "linter-coffeelint": "false",
  "linter-csslint": "true"
}
```
I am neutral on this idea, I am personally not gonna use it but I know a lot of people who will
/cc @atom-community/linter-contributors

**Comment by joefitzgerald** [comment]
Why wouldn't the linters just only run when they see the appropriate coffeelint.json (or equivalent) file, otherwise remaining disabled?

**Comment by steelbrain** [comment]
Because in most of the linters, to avoid complexity, we don't really find `coffeelint.json` or `.jshintrc`, we just call the CLI linter passing the filePath as an arg and it finds the config files automatically. Also their configs sometime contain quite complex stuff which wouldn't be ideal to parse ourselves.

**Comment by bluedaniel** [comment]
You don't have to parse the config file, you just need to check the existence of the file.

See how ESLint does it https://github.com/AtomLinter/linter-eslint/blob/master/lib/linter-eslint.coffee#L62

**Comment by steelbrain** [comment]
@joefitzgerald should I just close this idea as invalid or wait for voters?

**Comment by AsaAyers** [comment]
CoffeeLint comes with a set of defaults, so you don't have to configure it to use it. 

Having linter look for the config means we have to understand every possible place the configuration might live. CoffeeLint comes with it's own method of finding its configuration, which doesn't have to live in `coffeelint.json` in the root. It can also be embedded inside your `package.json` or `$HOME/coffeelint.json` (or whatever your OS's equivalent is). eslint and jshint also allow you to embed your config in `package.json`.

I think this is the best solution to this problem I've seen.

**Comment by bluedaniel** [comment]
Yes but, quite simply, doesn't `disableWhenNoJshlintrcFileInPath` do exactly what it says on the tin?

I do want some degree of control using this plugin and I'm not sure you guys are getting that.

**Comment by steelbrain** [comment]
@bluedaniel We are getting that :) but it isn't really ideal to scan all the paths that package itself is gonna check and then return empty errors if none is found, it would mean duplicate code in a lot of places, that we'll have to keep in sync with upstream.

**Comment by stevemao** [comment]
> The reason is that there are too much possibilities for where the config could be, it could be in your $HOME/.somethingrc, or embedded in your package.json or in your Project root. 
Copy/pasting all the logic from all the linters would increase the effort it takes to maintain a linter.

I wonder if, say jshint, has an api that returns the config (without merging with the defaults). If so, you could just call that instead of using the copy/pasted logic.

**Comment by steelbrain** [comment]
@stevemao I agree that a few of them provide such helper methods, but doing this operation in the main thread can cause a significant lag for the user, and if we do that on each `onDidStopChanging` event, it's gonna lag so much that people would want to disable it.
The method I mentioned above is the most efficient way to do it, if you have any other in mind, please do tell us :)

**Comment by stevemao** [comment]
Sorry I'm not quite following..
> The method I mentioned above is the most efficient way to do it

The method is "Copy/pasting all the logic from all the linters"?

So you can run those "Copy/pasting all the logic from all the linters" from background but you have to run linter helper methods in the main thread?

**Comment by steelbrain** [comment]
@stevemao The method is to allow the users to enable/disable linters using a JSON file.

**Comment by stevemao** [comment]
Cool. Thanks to clarify this @steelbrain :)

**Comment by stevemao** [comment]
I'm not sure if anyone has mentioned this but a list of linters somewhere in the editor that can be toggled for a project would be great. The list should be easily accessed and should save the result to the JSON file that @steelbrain mentioned. I don't know if the JSON should be in the project because some people might not use atom-linter in the project.

**Comment by AsaAyers** [comment]
> I don't know if the JSON should be in the project because some people might not use atom-linter in the project.

The project is definitely the right place to configure this. It doesn't matter if some members of your team don't use Atom. They can just ignore the one extra file they don't need.

**Comment by bluedaniel** [comment]
No offence but I don't want an editor plugin polluting my project files. You shouldn't be going this route (what other editor does this?).

**Comment by steelbrain** [comment]
@bluedaniel You can always git ignore, besides that's how all the packages that needs this kind-of configuration do it, `.jshintrc`, `.eslintrc`, `.babalrc`and a lot of others.

**Comment by bluedaniel** [comment]
Exactly, those are not atom packages but config files that can be used with any text editors on any platform.

Your suggesting an Atom specific configuration file in my project root, thats the difference.

**Comment by steelbrain** [comment]
@bluedaniel If you have a better alternative in mind, please do tell.

**Comment by AsaAyers** [comment]
After giving this some more thought I have revised my opinion.

![3pgxnw](https://cloud.githubusercontent.com/assets/324999/9155057/39491b6e-3e79-11e5-8965-4811ceaaed40.jpg)

`linter-coffeelint` doesn't allow you to configure it inside Atom because your configuration belongs in your project, never the editor.

Having re-read what this issue is about I realized that the configuration here is very specific to you. Maybe I have `linter-jshint` and `linter-eslint`, but our project is `linter-eslint`. Other members of my team don't need to disable linters they don't have installed.

**Comment by stevemao** [comment]
I tend to believe the JSON file should be a personal thing. Maybe this file should cascade.

**Comment by bluedaniel** [comment]
@steelbrain I really don't see the issue with each linter responsible for finding it's own config file and then disabling itself if told to do so in atom-config.

Its a method which already works well.

**Comment by steelbrain** [comment]
@bluedaniel We are at a stage with linter where performance is a critical priority, and can not afford anything like it. If we don't want it in JSON file, We could create a configuration page for it. But having this in each linter is a performance kill.

**Comment by bluedaniel** [comment]
@steelbrain A performance kill in startup time or linting on the fly?

I don't notice any performance issues with the 3 linter packages I use (scss, jshint, eslint)

**Comment by steelbrain** [comment]
@bluedaniel in linting, both on fly and on save. Everytime you save, the linter goes through the tree as well as some predefined locations to do some `fs.statSync` and `fs.readDirSync`, while that sync operation is being performed, it blocks the event loop, which means the editor would be unresponsive until that function has finished calling, If you are 5 dir deep, that means 5 scan dirs and 1 fs stat, if all of them take 5ms each, it means about 30ms per linter, per lint call. You can not PR on **all** of those linters to make it async, and even if you make it async, that still wouldn't mean that it's the best way to achieve that purpose, `linter-eslint` lagged my Atom and reduced it's fps to about 10 on  non-SSD 4th gen laptop, so I had to split the linting to a separate process. https://github.com/AtomLinter/linter-eslint/issues/184

**Comment by stevemao** [comment]
@steelbrain linters will search for config files anyway no matter what we do here right? they should not 1) search for config files, 2) lint (which search for config files again) but 1) search for config files, 2) if nothing is found don't do anything. So this might gain a bit of performance?

just a side note I'm not sure if this module can enforce it but all linters should not perform any blocking operation.

**Comment by steelbrain** [comment]
@stevemao when we're gonna import a linter, and use it's find config methods, they **are** gonna be sync and there's nothing we can do about them except a PR.
Those linters are gonna search for configs, yes. But they do it in a separate process, not in the main thread.

**Comment by AsaAyers** [comment]
This didn't start as a performance issue. The original reason was that
people like me work on multiple projects. Some use eslint and some use
jshint. It helps to be able to turn linters off per project.

I'm aware of the awkward disableIfMyConfigFileWasNotFound options. We
should still be able to turn the linter off instead.

**Comment by okonomiyaki3000** [comment]
I don't think it's that big of an issue. If people need to switch specific linters on and off on a project-by-project basis, why not use the [project-manager](https://atom.io/packages/project-manager) plugin? Not only will it allow switching them on and off, it let's you configure them differently for each project. I think there's no need to add this complexity to Linter when the functionality is already available. 

To put it another way, Linter doesn't need to be aware of projects since there's already a project-centric plugin that is aware of plugin settings in a general sense.

**Comment by polarathene** [comment]
Agree with @okonomiyaki3000 , any config related to specific editor/IDE is usually personal preference and stored within the project as a dot file or hidden directory. You should avoid the approach of a json file just to handle linter use for projects, Atom packages as a whole that would want to provide similar features should agree on a common approach (if separate files needed then they should all be stored within a `.atom_settings` folder or similar). The project-manager plugin looks like a great solution.

**Comment by steelbrain** [comment]
Okay people, here's an idea, would you like a similar UI for toggling providers?
![screenshot from 2015-10-13 14-43-55](https://cloud.githubusercontent.com/assets/4278113/10559219/8d95034e-74a8-11e5-9e05-6d0b7c173c45.png)

**Comment by johannes-z** [comment]
I'd consider linting options more a project related preference rather than an editor/IDE preference. Consider working on a project - everybody should be using the same options especially if it regards coding guidelines and styles.

**Comment by okonomiyaki3000** [comment]
@johannes-z That's true but that is not at issue here. Yes, your project needs to have dot files or whatever to define code style and other linting options. Your project does not need to have a file telling your ide which linting plugins to activate and which ones to disable.

Suppose a project has a file that lists which linters to use. You join this project and open it up for the first time. Some linters are automatically activated. Are others automatically deactivated? What if you haven't installed all the linter plugins? Are they installed for you automatically? What if you're using a different ide? What if you prefer a different plugin for one of the linters that the project uses? 

Which linters to use and how they are configured is a project issue. How to interact with those linters is a matter of developer preference and may not involve ide plugins at all.

**Comment by johannes-z** [comment]
@okonomiyaki3000 I thought this was an atom-centric issue and thus didn't consider differing IDEs / editors. This is indeed a problem since you can't ensure the use of linters in a project anyway. Considering a one-man project, it's very much a developer preference as you already mentioned. I for one would like to have different options/linters and settings on a project basis as you, but I can see the point in having a global definition which can be overriden by more specific options, namely project based configurations. 

While you have a point in the project not needing to have a file that tells me which linters to use, I would very much appreciate the file (or anything else defined in the project) to actually tell me, with which linters it has been set up with. 
This brings me back to the project-focused aspect. If I share my atom project with anyone who has a clean install of atom I want him to know which linters are being used in the project... Just a thought.

**Comment by florianb** [comment]
:100:

I'd like to encourage a decision on this topic for the following reason:

I know at least about two linter-plugins which definitely need a way to allow project-based linter-settings, which are [linter-javac](https://github.com/AtomLinter/linter-javac) and [linter-scalac](https://github.com/AtomLinter/linter-scalac).

In absence of a standardized possibility to configure the linters individually via config-files there were `.classpath`-files introduced. Unfortunately, the `.classpath`-file has its origin in the Eclipse-foundation and has at least in the Java-community a different syntax as the current implementation.

In my opinion it is absolutely necessary to find a unobtrusive way for project-based configurations.

**I therefore propose the introduction of an optional linter-wide configuration.**

**Comment by Arcanemagus** [comment]
How about a format like this?

```json
{
  "linter-foo": {
    "enable": true,
    "options": {
      "option1": "asdf",
      "option2": "fdsa"
    }
  },
  "linter-bar": {
    "enable": false
  }
}
```

At its most basic it provides the simple "enable/disable" discussed at the start, with the expanded capability of providing per project options to a linter provider. We could enhance the Linter API to provide a second (optional) argument to `lint()` with the set of options defined here.

**Comment by florianb** [comment]
@Arcanemagus: :+1:

I'd also like to add the following points to discussion:

1. Every setting is optional.
2. Every setting overrides the matching config-setting of the specific linter.
3. A setting-file will be searched at the project-roots and is in effect for the whole corresponding project.

As enhancement i could also imagine that:

a. setting-files will be searched from directory of the opened file downwards to the project-root, where every additional setting-file may set any absent setting.

To make integration easy, i'd suggest to let us try to figure out if it is possible to make `atom-linter` silently and reliable map the settings to the specific linter-config.

**Comment by Arcanemagus** [comment]
I don't think this should have anything to do with `atom-linter`, that's just a helper library for dealing with things like executing processes. Abstracting it out to that level would be rather annoying in any case as it runs in the context of the linter, but would need to control aspects outside that process.

If a linter provider wants to implement custom options, it should know what those are and how to handle them.

As for where setting files are located since we know this is going to be Atom specific I think requiring it to be at the project root, and _only_ there is acceptable, do we really need the ability to change settings per file/subfolder within a project?

**Comment by okonomiyaki3000** [comment]
:-1: Users who need this level of control should just install [project manager](https://atom.io/packages/project-manager). It already has a way to handle this stuff and, since it applies to all plugins, it's more useful than a feature that would only apply to linters.

Maybe there's something a little uncomfortable about telling linter users that they need a completely unrelated plugin to get a feature that (in their minds, at least) seems linter-centric. Well, I don't think that's a problem at all. I'd rather run a lot of plugins that each have a specific and focused role than a few that each do a lot of things (some of which will likely overlap). I guess that's the Unix philosophy but it should apply here too.

**Comment by Arcanemagus** [comment]
Oooo, didn't know that existed. I'm definitely with @okonomiyaki3000 here that we should just leave this up to that since it already exists (and from the look of it is quite popular).

Thanks for the heads up!

**Comment by florianb** [comment]
I don't like introducing a hard plugin-dependency for all java-linter-users but i'd support it if that's what the linter-dev-group wants.

**Comment by hd-deman** [comment]
Just a note: it is possible to use per project settings for plugins with [project-manager](https://atom.io/packages/project-manager) package.

**Comment by okonomiyaki3000** [comment]
@florianb it wouldn't be for all of them. It would be for the java-linter users who occasionally don't use java-linter. Among them, many may already be using project manager anyway.

**Comment by florianb** [comment]
They may or may not. That's outside my scope.

**Comment by Arcanemagus** [comment]
Marking this as closed since [project manager](https://atom.io/packages/project-manager) fullfills everything asked for so far in this issue (and then some).

@florianb This isn't a hard dependency, you just need to provide an option to be toggled by it, and tell users that haven't discovered it yet about it. Better than writing and maintaining that code yourself, no? :stuck_out_tongue:

If you have more to add or think this should be reopened feel free to keep commenting :wink:
